oeps/oep-0005.rst

```diff
@@ -0,0 +1,182 @@
+   makes it hard to update (and to push updates out to users).
+
+
+Specification
+=============
+
+Interface
+~~~~~~~~~
+
+The primary developer interaction with ``devstack`` will be through ``make``
+shortcuts. The following targets will be standard::
+
+    make devstack.watch  # Start the devstack services (if they aren't already)
+                         # and display their logs on stdout
+
+    make devstack.ssh    # Start the devstack services (if they aren't already)
+                         # and ssh into the primary service (as the user that
+                         # owns the service, in the directory that has the
+                         # service source code)
+
+    make devstack.clean  # Stop the devstack services and remove any intermediate
+                         # data that they stored in the .devstack/ directory
+
+    make devstack.*      # Start the devstack services (if they aren't already)
+                         # and run `make *` from the code directory of the service
+                         # (as the service owner)
+
+Implementation
+~~~~~~~~~~~~~~
+
+Requirements
+------------
+
+1. Services that aren't part of the Open edX ecosystem must be run using
+   their own published containers (where available).
+2. Open edX containers must run all of their processes via `supervisor`_ to
+   provide a single entrypoint even when the container has multiple processes
+   (such as `nginx`_ and a `Django`_ service running under `gunicorn`_).
+3. Open edX services running in containers must log to ``stdout`` and ``stderr``
+   so that those logs can be captured by `supervisor`_.
+4. Dependent services (such as databases) should be configured to mount their
+   storage directories in ``.devstack/`` inside the source code directory (this
+   should be added to `.gitignore` and `.dockerignore`_. This will allow data from
+   ``devstack`` to be persist across container executions.
+5. The Open edX service must mount the source code directory from the host
+   machine into the ``devstack`` container so that ``devstack`` is always running
+   the version of code being edited.
+6. ``devstack`` services should be configured to auto-reload when source code
+   is modified.
+7. Open edX Service containers should be built during CI, and published so that any
+   developer can download the latest container. They should also be tagged for
+   each versioned release.
+
+Technology Selection
+--------------------
+
+The prototype implementation was done using `Docker`_ for the containers, and
+`Docker Compose`_ to manage the networking between services.
+
+Rationale
+=========
+
+Choice of `Docker`_ over `Vagrant`_
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+While `Vagrant`_ has worked relatively well for the current monolithic Devstack
+implementation, it would be too resource heavy for running many services simultaneously.
+`Docker`_ has a much lower per-service overhead.
```

Comment:
Is that really so blocking? With one VM per service, each of the services VMs would have a smaller memory and CPU footprint, not all services would have to be run at the same time, and RAM is pretty cheap nowadays. I would be interested in looking at the actual figures - do we know what resources it would actually take in both cases? I assume you are comparing it to Vagrant multi-machine setups, ie https://www.vagrantup.com/docs/multi-machine/ ? Converting to docker has a high development time price tag, in addition to having to deal with its drawbacks mentioned above (single process orientation & pseudo virtualization).